### PR TITLE
swarm/hash: New package + custom multihash

### PR DESCRIPTION
**Pull request by nolash**
This PR addresses https://github.com/ethersphere/go-ethereum/issues/595 and https://github.com/ethersphere/go-ethereum/issues/427

It introduces the `swarm/hash` package, which contains a singleton collection of hashing handlers `Hasher` that use hash functions in a `sync.Pool` as backend. Hash functions are added by associating them with string identifiers. A hash function must be set as default before commencing operations, or a panic will occur. 

Client code interacts with the hash functions through the struct `Hasher`, which exposes the following methods:
* `Hash(data ...[]byte)` - resets and sequentially writes `...[]byte` to the hash function and returns the digets
* `HashWithLength(length []byte, data ...[]byte)` - performs `Hash.ResetWithLength(length)` before sequentially writing `...[]byte` to the hash function and returns the digest.
* `Size()` - returns the digest size of the hash function abstracted by the handler

The package members Hash and HashFunc replace the earlier `storage.SwarmHash` and `storage.SwarmHasher` constructs.

-----

The custom multihash implementation aims to be as thin as possible, and currently makes the assumption that _only the default hash function will be used in the context of multihash_.

### CODE REVIEW

File: swarm/hash/types.go [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+)
+
+// the basic swarm hash interface
+// an instance represents an individual worker in a pool
+type Hash interface {
+	hash.Hash
+	ResetWithLength([]byte)
+}
+
+// function for creating a new individual worker
```
Comment (zelig) [inline swarm/hash/types.go:42]: if you define this, use it on line 50

File: swarm/hash/types.go [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+}
+
+func init() {
+	pools = make(map[string]*Hasher)
+}
+
+// Init sets the default hash to be used when calling GetHash
+// This must be executed before any hashing can commence
+// If no hash functions is registered with the identifier, the function will panic
+func Init(typ string) {
```
Comment (zelig) [inline swarm/hash/types.go:62]: this is quite ugly to initialise from a global default const
let make a contructor then, and pass argument in the code.

File: swarm/hash/types.go [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+// HashWithLength does the same as Hash, but first calls ResetWithLength with the provided "length" parameter on the underlying hash
+func (h *Hasher) HashWithLength(length []byte, data ...[]byte) []byte {
+	hasher := h.pool.Get().(Hash)
+	defer h.pool.Put(hasher)
+	hasher.ResetWithLength(length)
+	return h.doHash(hasher, data)
+}
+
+// backend function for Hash*()
+func (h *Hasher) doHash(hasher Hash, data [][]byte) []byte {
```
Comment (zelig) [inline swarm/hash/types.go:159]: why data is not `data ...[]byte` here and is in `HashWithLength`?

File: swarm/storage/chunker_test.go [hunk 4]
```diff
@@ -62,7 +62,7 @@ func (f *fakeChunkStore) Close() {
 }
 
 func newTestHasherStore(chunkStore ChunkStore, hash string) *hasherStore {
-	return NewHasherStore(chunkStore, MakeHashFunc(hash), false)
+	return NewHasherStore(chunkStore, false)
```
Comment (zelig) [inline swarm/storage/chunker_test.go:65]: I still think the hasher should be passed, even if it is a global service

File: swarm/storage/resource_test.go [hunk 5]
```diff
@@ -423,23 +411,12 @@ func TestResourceMultihash(t *testing.T) {
...
-		t.Fatal(err)
-	}
-	sha1decode, err := multihash.Decode(data)
+	swarmhashdecode, err := swarmhash.FromMultihash(data)
 	if err != nil {
 		t.Fatal(err)
 	}
-	if !bytes.Equal(sha1decode.Digest, sha1bytes) {
-		t.Fatalf("Decoded SHA1 hash '%x' does not match original hash '%x'", sha1decode.Digest, sha1bytes)
+	if !bytes.Equal(swarmhashdecode, swarmhashbytes.Bytes()) {
```
Comment (zelig) [inline swarm/storage/resource_test.go:418]: sha1?

File: swarm/storage/types.go [hunk 6]
```diff
@@ -22,21 +22,23 @@ import (
...
 	"github.com/ethereum/go-ethereum/log"
+	swarmhash "github.com/ethereum/go-ethereum/swarm/hash"
 )
 
 const MaxPO = 16
 const KeyLength = 32
 
-type Hasher func() hash.Hash
-type SwarmHasher func() SwarmHash
+func init() {
```
Comment (zelig) [inline swarm/storage/types.go:38]: again i think this should be a proper constructor

File: swarm/storage/types.go [hunk 7]
```diff
@@ -105,14 +107,14 @@ func IsZeroKey(key Key) bool {
...
 	switch hash {
 	case "SHA256":
-		return func() SwarmHash { return &HashWithLength{crypto.SHA256.New()} }
+		return func() swarmhash.Hash { return &HashWithLength{crypto.SHA256.New()} }
 	case "SHA3":
-		return func() SwarmHash { return &HashWithLength{sha3.NewKeccak256()} }
+		return func() swarmhash.Hash { return &HashWithLength{sha3.NewKeccak256()} }
 	case "BMT":
-		return func() SwarmHash {
+		return func() swarmhash.Hash {
```
Comment (zelig) [inline swarm/storage/types.go:117]: if anything the BMT should use the SHA3 hashing service. BMT itself not sure if benefits from this worker thing. have you checked somehow?

File: swarm/storage/types.go [hunk 8]
```diff
@@ -328,22 +325,17 @@ type ChunkValidator interface {
...
-	Hasher SwarmHasher
 }
 
 // Constructor
-func NewContentAddressValidator(hasher SwarmHasher) *ContentAddressValidator {
-	return &ContentAddressValidator{
-		Hasher: hasher,
-	}
+func NewContentAddressValidator() *ContentAddressValidator {
+	return &ContentAddressValidator{}
```
Comment (zelig) [inline swarm/storage/types.go:332]: again should be passed as arg

**Comment by nolash** [comment]
Maybe we should talk through this @zelig. Maybe with @gbalint  too?

File: swarm/storage/types.go (hunk 7 above)
Comment (nolash) [reply swarm/storage/types.go:117]: Does the benchmark mentioned in the PR text cover your question of benefit?

https://github.com/ethersphere/go-ethereum/issues/427

**Comment by nolash** [comment]
Obsoleted by https://github.com/ethersphere/go-ethereum/pull/633 and https://github.com/ethersphere/go-ethereum/pull/634

**Pull request closed without merging**